[LLDB][Docs] List available settings

Nerixyz opened the pull request:
This PR adds a documentation page that lists all available settings. The page is automatically generated.

You can find a preview at https://nerixyz.github.io/test-gh-pages/use/settings.html

Having the settings listed in the online documentation makes it easier to search for users. It also has the advantage of being indexed by search engines.

To generate the docs, we first generate JSON out of the TableGen definitions with `-dump-json`.

Once all properties are available as JSON, a Markdown file with the merged documentation (`settings.md`) is generated. I chose Markdown over RST, because some descriptions already use backticks, which would become references in RST.

Currently, enum names/descriptions are not visible, because they're defined in C++. In the future, these could be moved to TableGen as well.

medismailben reviewed (approved):
Pretty cool! I was a bit skeptical to put the generator command in `lldb-test`. Why not use the main `lldb` binary for this ?

JDevlieghere reviewed (changes requested):
I'm very supportive of the idea, but I have concerns with the implementation.

You say that currently that building the documentation depends on `liblldb` but I don't believe that's accurate (or if it is, that's a regression). The only thing we should depend on is on the SWIG wrapper (`lldb.py`) which only requires parsing the API headers. 

Could we achieve the same thing by parsing the TableGen for the settings directly? I did [something similar](https://github.com/llvm/llvm-project/blob/main/llvm/utils/TableGen/OptionRSTEmitter.cpp) for generating the command line options.

DavidSpickett commented:
> You say that currently that building the documentation depends on liblldb but I don't believe that's accurate (or if it is, that's a regression). The only thing we should depend on is on the SWIG wrapper (lldb.py) which only requires parsing the API headers.

I've seen people claim this before.

https://github.com/llvm/llvm-project/issues/123316 is along the same lines.

Nerixyz commented:
> You say that currently that building the documentation depends on liblldb but I don't believe that's accurate (or if it is, that's a regression). 

I only looked at the output of the documentation builder (e.g. https://lab.llvm.org/buildbot/#/builders/45/builds/18201/steps/5/logs/stdio). This shows that liblldb is indeed built.

I'll try to do the generation with just the Tablegen files. Though I would like to have something similar for commands (a page the shows all available ones and their options). There, we'd need to convert alltheir definitions to Tablegen, I suppose.

Nerixyz commented:
> Could we achieve the same thing by parsing the TableGen for the settings directly?

At least not without adding more (potentially) duplicate information about the settings hierarchy. From the TableGen files alone, we don't know the parents.
Take the C++ language plugin as an example. In the plugin properties (https://github.com/llvm/llvm-project/blob/3e8dc4dc4d04fe4c42f139423a61802b1ba719fc/lldb/source/Plugins/Language/CPlusPlus/LanguageCPlusPlusProperties.td), one setting (`unction-name-format`) is declared. Going just by the location in the tree, it could be in `plugin.cplusplus`. However, it's actually in `plugin.cplusplus.display`.

Nerixyz committed:
[LLDB][Docs] Generate documentation for settings

DavidSpickett commented on lldb/utils/TableGen/LLDBTableGen.cpp:
```diff
@@ -41,7 +42,9 @@ static cl::opt<ActionType> Action(
-                          "Generate lldb property enum definitions")));
+                          "Generate lldb property enum definitions"),
+               clEnumValN(GenPropertyDocsJson, "gen-lldb-property-docs-json",
+                          "Generate lldb property documentation")));
```
Add "in JSON format" here. I see what you mean but my mind immediately jumps to markdown or html, some document format.

Nerixyz commented:
> Convenience I assume, but how much so?

Yes. Though looking at the `-dump-json` output now, I'm thinking it's not that complicated to use. I'll give it a try.

DavidSpickett commented on lldb/source/Plugins/Process/MacOSX-Kernel/CMakeLists.txt:
```diff
@@ -6,6 +6,14 @@ lldb_tablegen(ProcessKDPPropertiesEnum.inc -gen-lldb-property-enum-defs
+
+if (NOT CMAKE_SYSTEM_NAME MATCHES "Darwin")
+  return()
+endif()
```
Yes, I see it in Plugins/Process/CMakeLists.txt.

Nerixyz committed:
refactor: use `-dump-json`

Nerixyz committed:
fix: use chained `:not`

Nerixyz committed:
fix: add note about LLDB configurations

Nerixyz committed:
fix: mention directive file

Nerixyz committed:
refactor: generate settings in build directory

Nerixyz committed:
fix: pass env with space

Nerixyz committed:
fix: only add directive if env variable is present

Nerixyz committed:
fix: dependency and levels

Nerixyz committed:
chore: remove now unused tablegen backend

DavidSpickett reviewed (approved):
I can't say whether the BuildInclude could be simpler or is reinventing a wheel, but what it does is not complex so it can always be replaced later.

So LGTM but wait for @JDevlieghere to confirm that they're happy with it.

DavidSpickett commented:
I know they are at LLVM Dev this week but ping @JDevlieghere.

I'd very much like to see this on the website.

Nerixyz committed:
Merge remote-tracking branch 'upstream/main' into docs/lldb-settings

Nerixyz committed:
fix: include freebsd-kernel-core settings

Nerixyz commented:
Updated to main and synced the preview (https://nerixyz.github.io/test-gh-pages/use/settings.html). For the freebsd-kernel-core settings, I changed the CMake setup slightly to ensure we always dump the JSON from tablegen for the settings (CC @mchoo7).

mchoo7 reviewed (approved):
LGTM!

Nerixyz committed:
fix: stylesheet for s/uint64

JDevlieghere commented on lldb/docs/CMakeLists.txt:
```diff
@@ -37,6 +37,22 @@ if (LLDB_ENABLE_PYTHON AND SPHINX_FOUND)
+      OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/generated-settings.md"
+      COMMAND 
+        "${Python3_EXECUTABLE}"
+        "${CMAKE_CURRENT_SOURCE_DIR}/_scripts/gen-property-docs-from-json.py"
```
Can this script live in the top level (lldb) scripts for?

Nerixyz replied:
Done.

JDevlieghere commented on lldb/docs/CMakeLists.txt:
```diff
@@ -46,13 +62,17 @@ if (LLDB_ENABLE_PYTHON AND SPHINX_FOUND)
     # the path to the module via the LLDB_SWIG_MODULE environment variable.
-    add_sphinx_target(html lldb ENV_VARS "LLDB_SWIG_MODULE=${CMAKE_CURRENT_BINARY_DIR}")
+    add_sphinx_target(html lldb
+      ENV_VARS
```
Are environment variables the only way to pass arguments?

Nerixyz replied:
At least the [manpage](https://www.sphinx-doc.org/en/master/man/sphinx-build.html) doesn't list anything else. `-D` can be used to change a setting _after_ `conf.py` runs (so it overwrites it), but we need the argument inside the `conf.py`.

JDevlieghere commented on lldb/source/Plugins/Process/FreeBSD-Kernel-Core/CMakeLists.txt:
```diff
@@ -6,6 +6,14 @@ lldb_tablegen(ProcessFreeBSDKernelCorePropertiesEnum.inc -gen-lldb-property-enum
+  TARGET LLDBPluginProcessFreeBSDKernelCorePropertiesJsonGen)
+
+if (NOT CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
+  return()
```
As much as I like early returns, I don't remember seeing this pattern elsewhere in llvm. Unless I just missed it, I would take the indentation of wrapping the rest of the file in an if block.

Nerixyz replied:
Done in both MacOSX-Kernel and FreeBSD-Kernel-Core.

JDevlieghere reviewed (approved):
I did a quick pass on my phone. I didn't see anything worth holding this up for. We can iterate post commit. LGTM!

Nerixyz committed:
fix: move script to lldb/scripts

Nerixyz committed:
fix: invert if with early return

Nerixyz committed:
fix: update other script location

medismailben commented:
Very cool! It's already live https://lldb.llvm.org/use/settings.html
